packages/core/src/combinator/zipItems.js

```diff
@@ -1,16 +1,14 @@
 /** @license MIT License (c) copyright 2017 original author or authors */
 
-import { empty } from '../source/empty'
-import { take } from './slice'
 import Pipe from '../sink/Pipe'
+import { isCanonicalEmpty } from '../source/empty'
+import { take } from './slice'
 
 export const withItems = (items, stream) =>
   zipItems(keepLeft, items, stream)
 
 export const zipItems = (f, items, stream) =>
-  items.length === 0 || stream === empty()
-    ? empty()
-    : new ZipItems(f, items, take(items.length, stream))
+  isCanonicalEmpty(stream) ? stream : new ZipItems(f, items, take(items.length, stream))
```

Comment:
I think we're missing an opportunity to propagate empty() in the case where items.length === 0.  In that case, rather than returning empty(), this will still return a ZipItems stream.  That ZipItems will _contain_ an empty(), but will prevent any further empty() propagation.

I see two ways to handle it.  First, re-add the other check and explicitly return empty():

```js
isCanonicalEmpty(stream) || items.length === 0 ? empty() : new ZipItems(f, items, take(items.length, stream))
```

Second, flip take and ZipItems.  This reads nicely, but is slightly more wasteful, because it constructs the ZipItems only to throw it away:

```js
isCanonicalEmpty(stream) ? stream : take(items.length, new ZipItems(f, items, stream))
```

In the grand scheme, unless you're calling zipItems in a tight loop with an empty array, you'll never notice the extra ZipItems construction.

Thoughts?

Reply:
> This reads nicely, but is slightly more wasteful ...

What kind of waste are we talking here compared to returning a `ZipItems` instance?

Reply:
i like the 1st ugly one

Reply:
Yeah, I see this a good solution
```typescript
isCanonicalEmpty(stream) ? stream : take(items.length, new ZipItems(f, items, stream))
```

Reply:
The first accesses the `Array.prototype.length` property twice.

Reply:
Updated with the second

Reply:
> What kind of waste are we talking here compared to returning a ZipItems instance?

I just meant that in the second version, if items.length === 0, it constructs a new ZipItems() and passes it to `take`, which take will then discard and return `empty()`.  So, the waste is the cost of allocation the unnecessary ZipItems().  A memory allocation like that is typically "more expensive" (in memory and cpu) than the second `items.length` access.  But in the grand scheme, the difference is probably tiny.  I doubt it will matter and readability seems more important.

Here's another possible approach that avoids both the double items.length access and the unnecessary ZipItems allocation:

```js
export const zipItems = (f, items, stream) =>
  privateZipItems(f, items, take(items.length, stream))

const privateZipItems = (f, items, stream) =>
  isCanonicalEmpty(stream) ? stream : new ZipItems(f, items, stream)
```

That relies on `take` to return empty() in the cases where items.length === 0 *or* stream is empty().  IOW, it relies on `take` to fold the two cases into one, so that `privateZipItems` only needs to check `isCanonicalEmpty`.

How does that one compare?

Reply:
I like that :heart: Just, I'd call it `internalZipItems` instead of _private_.

Reply:
Yeah, I'm good with option 1 or 3.

Reply:
It's no. 1 for me.